### PR TITLE
[MAINT] add docker GHA worklow

### LINKED ISSUE #69: test the build of the docker image more regularly
Probably want to automate testing "just" the docker build step on regular basis: from experience it helps catching issues when dependencies get updated / deprecated.

We could also have the workflow be triggered on PR that touch specific files (Dockerfile for example).

### PR DESCRIPTION
**Pull request by Remi-Gau**
- closes #69

- move docker tests to a separate workflow

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -160,64 +160,3 @@ jobs:
...
-      - name: Upload output artifact
-        uses: actions/upload-artifact@v4
-        with:
-          name: connectome
-          path: ./outputs/
-      - name: Upload docker image artifact
-        uses: actions/upload-artifact@v4
-        with:
-          name: docker-image
-          path: ${HOME}/docker/image.tar
```
Comment (Remi-Gau) [inline .github/workflows/test.yml:209]: probably worth transferring to the other workflow to be able to inspect the output

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [submodule "tools/mist2templateflow"]
 	path = tools/mist2templateflow
-	url = https://github.com/SIMEXP/mist2templateflow.git
```
Comment (Remi-Gau) [inline .gitmodules:3]: temporary switching to my fork to test with the fix for the nilearn version

File: pyproject.toml [hunk 3]
```diff
@@ -20,8 +20,7 @@ classifiers = [
   "Programming Language :: Python :: 3.12",
 ]
 dependencies = [
-  "h5py",
-  "nilearn[plotting] >=0.10.2",
+  "nilearn[plotting] >=0.10.3",
```
Comment (Remi-Gau) [inline pyproject.toml:23]: need to bump nilearn version to be consistent with that of mist2templateflow

File: pyproject.toml [hunk 4]
```diff
@@ -20,8 +20,7 @@ classifiers = [
   "Programming Language :: Python :: 3.12",
 ]
 dependencies = [
-  "h5py",
```
Comment (Remi-Gau) [inline pyproject.toml:23]: forgot to remove it

**Comment by Remi-Gau** [comment]
Getting the following error when running in docker:

```bash
  docker run --rm --read-only \
    -v /home/runner/work/giga_connectome/giga_connectome/giga_connectome/data/test_data:/test_data \
    -v ./outputs:/outputs \
    -v ./outputs/working_dir:/work \
      bids/giga_connectome \
        /test_data/ds000017-fmriprep22.0.1-downsampled-nosurface \
        /outputs \
        participant \
        -w /work \
        --participant_label 1 \
        --reindex-bids

Traceback (most recent call last):
  File "messagestream.pyx", line 91, in scipy._lib.messagestream.MessageStream.close
TypeError: remove: path should be string, bytes or os.PathLike, not NoneType
Exception ignored in: 'scipy._lib.messagestream.MessageStream.__dealloc__'
Traceback (most recent call last):
  File "messagestream.pyx", line 91, in scipy._lib.messagestream.MessageStream.close
TypeError: remove: path should be string, bytes or os.PathLike, not NoneType
Traceback (most recent call last):
  File "/usr/local/bin/giga_connectome", line 5, in <module>
    from giga_connectome.run import main
  File "/usr/local/lib/python3.9/site-packages/giga_connectome/__init__.py", line 9, in <module>
    from .atlas import load_atlas_setting
  File "/usr/local/lib/python3.9/site-packages/giga_connectome/atlas.py", line 10, in <module>
    from nilearn.image import resample_to_img
  File "/usr/local/lib/python3.9/site-packages/nilearn/image/__init__.py", line 6, in <module>
    from .image import (
  File "/usr/local/lib/python3.9/site-packages/nilearn/image/image.py", line 18, in <module>
    from scipy.stats import scoreatpercentile
  File "/usr/local/lib/python3.9/site-packages/scipy/stats/__init__.py", line 623, in <module>
    from ._rvs_sampling import rvs_ratio_uniforms
  File "/usr/local/lib/python3.9/site-packages/scipy/stats/_rvs_sampling.py", line 2, in <module>
    from scipy.stats.sampling import RatioUniforms
  File "/usr/local/lib/python3.9/site-packages/scipy/stats/sampling.py", line 59, in <module>
    from ._sampling import FastGeneratorInversion, RatioUniforms  # noqa: F401
  File "/usr/local/lib/python3.9/site-packages/scipy/stats/_sampling.py", line 8, in <module>
    from ._unuran.unuran_wrapper import NumericalInversePolynomial
  File "unuran_wrapper.pyx", line 188, in init scipy.stats._unuran.unuran_wrapper
  File "unuran_wrapper.pyx", line 167, in scipy.stats._unuran.unuran_wrapper._setup_unuran
  File "messagestream.pyx", line 29, in scipy._lib.messagestream.MessageStream.__cinit__
  File "/usr/local/lib/python3.9/tempfile.py", line 329, in mkstemp
    prefix, suffix, dir, output_type = _sanitize_params(prefix, suffix, dir)
  File "/usr/local/lib/python3.9/tempfile.py", line 124, in _sanitize_params
    dir = gettempdirb()
  File "/usr/local/lib/python3.9/tempfile.py", line 298, in gettempdirb
    return _os.fsencode(gettempdir())
  File "/usr/local/lib/python3.9/tempfile.py", line 291, in gettempdir
    tempdir = _get_default_tempdir()
  File "/usr/local/lib/python3.9/tempfile.py", line 223, in _get_default_tempdir
    raise FileNotFoundError(_errno.ENOENT,
FileNotFoundError: [Errno 2] No usable temporary directory found in ['/tmp', '/var/tmp', '/usr/tmp', '/code']
```

**Comment by Remi-Gau** [comment]
hum...

Tried to reproduce the issue in a virtual env with the dependencies in the requirements.txt but I could not.

Only difference is that my env uses python 3.12 and the docker images uses python 3.9.

**Comment by htwangtw** [comment]
This is because the docker container is set to read-only in the GitHub action, I believe?
I am not sure what's a good work around.

File: .gitmodules (hunk 2 above)
Comment (htwangtw) [reply .gitmodules:3]: I have merged the version bump!

Comment (Remi-Gau) [reply .gitmodules:3]: will switch back top the upstream repo

**Comment by Remi-Gau** [comment]
> This is because the docker container is set to read-only in the GitHub action, I believe? I am not sure what's a good work around.

Good point. Let me check.

**Comment by Remi-Gau** [comment]
OK now hitting a `[Errno 28] No space left on device`

**Comment by Remi-Gau** [comment]
Ok good news the build and the run are working.

Now need to figure out the push.

**Comment by Remi-Gau** [comment]
Hum... Set the secrets at the orga level but this seems not to work. Will try to debug when not on a phone.

**Comment by htwangtw** [comment]
This is amazing thanks a bunch

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      - name: Restore cached docker image
+        uses: actions/cache/restore@v4
+        with:
+          path: ${{ env.IMAGE }}
+          key: docker
+      - name: Load image
+        run: docker load -i ${{ env.IMAGE }}/image.tar
+      - name: Push unstable to dockerhub on tags or on main
+        if:  ${{  github.ref_name }} == "main" || ${{  github.ref_type }} == "tag"
+        run: |
```
Comment (htwangtw) [inline .github/workflows/docker.yml:117]:
```suggestion
       env:
         DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
         DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
       run: |
```

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+          if [[ -n "${{ secrets.DOCKERHUB_TOKEN }}" ]]; then
```
Comment (htwangtw) [inline .github/workflows/docker.yml:118]:
```suggestion
          if [[ -n "$DOCKERHUB_TOKEN" ]]; then
```

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+            echo "${{ secrets.DOCKERHUB_TOKEN }}" | docker login -u "${{ secrets.DOCKERHUB_USERNAME }}" --password-stdin
```
Comment (htwangtw) [inline .github/workflows/docker.yml:120]:
```suggestion
            echo "$DOCKERHUB_TOKEN" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
```

File: .github/workflows/docker.yml [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            docker tag "${{env.USER_NAME}}/${{env.REPO_NAME}}" "${unstable}"
+            docker push "${unstable}"
+          else
+            : "No DOCKER_TOKEN, skipping push to DockerHub"
+            exit 1
+          fi
+      - name: Push stable release to dockerhub on tags only
+        if: ${{  github.ref_type }} == "tag"
+        run: |
+          if [[ -n "${{ secrets.DOCKERHUB_TOKEN }}" ]]; then
```
Comment (htwangtw) [inline .github/workflows/docker.yml:135]:
```suggestion
          if [[ -n "DOCKERHUB_TOKEN" ]]; then
```

File: .github/workflows/docker.yml [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+            echo "${{ secrets.DOCKERHUB_TOKEN }}" | docker login -u "${{ secrets.DOCKERHUB_USERNAME }}" --password-stdin
```
Comment (htwangtw) [inline .github/workflows/docker.yml:137]:
```suggestion
            echo "$DOCKERHUB_TOKEN" | docker login -u "$DOCKERHUB_USERNAME " --password-stdin
```

**Review by htwangtw — COMMENTED** [review]
Looks like this is the syntax. For reasons secrets has to be defined as a variable?

https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#example-using-bash

**Comment by Remi-Gau** [comment]
> Looks like this is the syntax. For reasons secrets has to be defined as a variable?

OK let's try

File: .github/workflows/docker.yml (hunk 5 above)
Comment (htwangtw) [reply .github/workflows/docker.yml]: my indentation might be wrong here

File: .github/workflows/docker.yml [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            docker tag "${{env.USER_NAME}}/${{env.REPO_NAME}}" "${unstable}"
+            docker push "${unstable}"
+          else
+            : "No DOCKER_TOKEN, skipping push to DockerHub"
+            exit 1
+          fi
+      - name: Push stable release to dockerhub on tags only
+        if: ${{  github.ref_type }} == "tag"
+        run: |
+          if [[ -n "DOCKERHUB_TOKEN" ]]; then
```
Comment (Remi-Gau) [inline .github/workflows/docker.yml:138]:
```suggestion
          if [[ -n "$DOCKERHUB_TOKEN" ]]; then
```

File: .github/workflows/docker.yml [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+          path: ${{ env.IMAGE }}
+          key: docker
+      - name: Load image
+        run: docker load -i ${{ env.IMAGE }}/image.tar
+      - name: Push unstable to dockerhub on tags or on main
+        if:  ${{  github.ref_name }} == "main" || ${{  github.ref_type }} == "tag"
+       env:
+         DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
+         DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
+       run: |
```
Comment (htwangtw) [inline .github/workflows/docker.yml:120]:
```suggestion
        env:
          DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
          DOCKERHUB_TOKEN: ${{ secrets.DOCKERHUB_TOKEN }}
        run: |
```

**Comment by htwangtw** [comment]
Let me check things...
This might be the real reason:

> Secrets cannot be directly referenced in if: conditionals. Instead, consider setting secrets as job-level environment variables, then referencing the environment variables to conditionally run steps in the job. For more information, see "[Contexts](https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability)" and [jobs.<job_id>.steps[*].if](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif).

File: .github/workflows/docker.yml [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            exit 1
+          fi
+      - name: Push stable release to dockerhub on tags only
+        if: ${{  github.ref_type }} == "tag"
+        run: |
+          if [[ -n "$DOCKERHUB_TOKEN" ]]; then
+
+            echo "$DOCKERHUB_TOKEN" | docker login -u "$DOCKERHUB_USERNAME " --password-stdin
+
+            : "Pushing stable and latest versions to DockerHub for latest and  ${{  github.ref_name }}"
```
Comment (Remi-Gau) [inline .github/workflows/docker.yml:142]:
```suggestion
            echo "Pushing stable and latest versions to DockerHub for latest and  ${{  github.ref_name }}"
```

File: .github/workflows/docker.yml [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+            latest="${{env.USER_NAME}}/${{env.REPO_NAME}}:latest"
+            docker tag "${unstable}" "${latest}"
+            docker push "${latest}"
+            tagged_release="${{env.USER_NAME}}/${{env.REPO_NAME}}:${{  github.ref_name }}"
+            docker tag "${unstable}" "${tagged_release}"
+            docker push "${tagged_release}"
+
+          else
+              : "No DOCKER_TOKEN, skipping push to DockerHub"
```
Comment (Remi-Gau) [inline .github/workflows/docker.yml:152]:
```suggestion
              echo "No DOCKER_TOKEN, skipping push to DockerHub"
```

**Comment by Remi-Gau** [comment]
closing in favor of #126

**Pull request closed without merging**